### PR TITLE
FINERACT-429: Node aware Scheduler

### PR DESCRIPTION
**Pull request by wkk91193**
[Fineract-429](https://issues.apache.org/jira/browse/FINERACT-429)

Executing jobs based on node id

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/scheduledjobs/service/ScheduledJobRunnerServiceImpl.java [hunk 1]
```diff
@@ -409,4 +416,16 @@ public void postDividends() throws JobExecutionException {
...
 
+    
+    //get all dirty jobs
+    // execute each job.
+    @Override
+    @CronTarget(jobName = JobName.EXECUTE_DIRTY_JOBS)
+    public void executeDirtyJobs() throws JobExecutionException {
+        List<ScheduledJobDetail> jobDetails=this.scheduledJobDetailsRepository.findAllDirtyJobs(true);
+        for (ScheduledJobDetail scheduledJobDetail:jobDetails){
+            jobRegisterService.executeJob(scheduledJobDetail.getId());
```
Comment (avikganguly01) [inline fineract-provider/src/main/java/org/apache/fineract/scheduledjobs/service/ScheduledJobRunnerServiceImpl.java:427]: It should execute only if that job node id and system node id matches.

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/jobs/service/JobRegisterServiceImpl.java [hunk 2]
```diff
@@ -199,7 +229,11 @@ public void startScheduler() {
                 for (final ScheduledJobDetail jobDetail : scheduledJobDetails) {
                     if (jobDetail.isTriggerMisfired()) {
                         if (jobDetail.isActiveSchedular()) {
-                            executeJob(jobDetail, SchedulerServiceConstants.TRIGGER_TYPE_CRON);
+                            if (jobDetail.getNodeId().toString()==System.getProperty("node.id")) {
```
Comment (avikganguly01) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/jobs/service/JobRegisterServiceImpl.java:232]: Also handle for a NodeID value which will run in all nodes like 0.

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/jobs/service/JobRegisterServiceImpl.java [hunk 3]
```diff
@@ -129,6 +135,30 @@ public void loadAllJobs() {
...
+            fis.close();
+        } catch (FileNotFoundException e) {
+            System.setProperty(nodeKey, defaultValue);
+        } catch (IOException e) {
+            System.setProperty(nodeKey, defaultValue);
+        }
+    }
+
     public void executeJob(final ScheduledJobDetail scheduledJobDetail, String triggerType) {
         try {
```
Comment (avikganguly01) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/jobs/service/JobRegisterServiceImpl.java:163]: I see only changes in executeJob function and not in schedule function.

**Comment by ShruthiRajaram** [comment]
@avikganguly01 kindly review this PR

**Comment by vorburger** [comment]
@wkk91193 thank you very much for your interest in and contribution to Fineract/Mifos!  I'm an (old) volunteer who got reminded of Mifos/Fineract this week-end at the FOSDEM conference, and I'm taking a moment to help the project clean up some of it's old issues and pull requests etc.  

We were wondering whether you are still alive, and would be willing to `git rebase develop` this PR and take any previous review feedback into account?  If we don't hear from you within 2 weeks, we will take the liberty to close this PR without merging it, to avoid confusion for future new contributors looking at this repository - hope you understand and that's fair.

/Cc FYI @apache/fineract-committers

**Review by avikganguly01 — CHANGES_REQUESTED** [review]
Please resolve the conflicts.

**Comment by vorburger** [comment]
@wkk91193 I'm currently doing a quick run of review on all open PRs, and was just wondering if you would still like to get this very old open one merged? If so, could you rebase it, so that we can see if it actually builds on Travis? (FYI when this PR was raised, this project did not have automated CI on Travis set up; but meanwhile it does.)

This one also has review feedback from feedback @avikganguly01 which needs to be taken into account.

Without hearing back within let's say ... 1 month, I would be inclined to just close this (un-merged).

**Comment by vorburger** [comment]
Let's close this now, just to have less open PRs; you can open a new PR with this when you resolved the feedback and conflicts.

**Pull request closed without merging**